notion-core/src/main/scala/com/ambiata/notion/core/PartReducer.scala

```diff
@@ -0,0 +1,69 @@
+package com.ambiata.notion.core
+
+import java.security.MessageDigest
+
+import com.ambiata.mundane.io.Checksum
+import org.apache.commons.codec.binary.Base64
+
+import scala.collection.mutable.Queue
+import scalaz.{Foldable, Monoid}
+
+/**
+ * "Reducer-like" class to accumulate values while
+ * reading "parts" (bytes, lines) from a file and transforming them to a final value
+ * when the whole file has been read
+ */
+trait PartReducer[P, T] {
+  type S
+  
+  /** initial value */
+  def init: S
+
+  /** reducer function using the current part and the current state S */
+  def reduce(part: P, s: S): S
+
+  /** final value from the last state */
+  def finalise(s: S): T
+}
+
+object PartReducer {
+
+  type LineReducer[T] = PartReducer[String, T]
+  type BytesReducer[T] = PartReducer[ReadBytes, T]
+
+  /** zip two line reducers together - aka parallel composition */
+  implicit class ZippedPartReducer[P, T1](r1: PartReducer[P, T1]) {
```

Comment:
I have to second Mark on this one, I'm not a fan of the implicit classes. It's something that we don't really use in Ivory and I find things easier to understand without having to track them down.

I even think we should put a note about them in the scala practices.